### PR TITLE
prov/sm2: Fix startup resource leaks

### PR DESCRIPTION
**Pull request by lrbison**
Fix CID 384050 by allowing common_fd=0 and catching mmap-related errors.

Fix similar leak when mkostemp file creation or map fails.

Change size of mkostemp file's template, as it's not related to FI_NAME_MAX.

check return values of several system calls which were previously ignored.

### CODE REVIEW

File: prov/sm2/src/sm2_coordination.c [hunk 1]
```diff
@@ -205,20 +205,20 @@ ssize_t sm2_file_open_or_create(struct sm2_mmap *map_shared)
...
-
-	sprintf(template, "%s/fi_sm2_pid%d_XXXXXX", SM2_COORDINATION_DIR,
+	sprintf(template, "%s/fi_sm2_pid%7d_XXXXXX", SM2_COORDINATION_DIR,
 		getpid());
 
 	/* Assume we are the first process here.
 	 * Try to open a tmpfile file in the shm directory
 	 */
 	fd = mkostemp(template, O_RDWR);
-	assert(fd > 0);
```
Comment (wenduwan) [inline prov/sm2/src/sm2_coordination.c:216]: Shouldn't we check fd immediately upon return?

File: prov/sm2/src/sm2_coordination.c [hunk 2]
```diff
@@ -192,7 +192,7 @@ ssize_t sm2_file_open_or_create(struct sm2_mmap *map_shared)
 {
 	pthread_mutexattr_t att;
 	struct sm2_mmap map_ours;
-	char template[FI_NAME_MAX];
+	char template[sizeof(SM2_COORDINATION_DIR) + 32];
```
Comment (a-szegel) [inline prov/sm2/src/sm2_coordination.c:195]: Not in love with the magic number, but its not the end of the world for me.

File: prov/sm2/src/sm2_coordination.c [hunk 3]
```diff
@@ -268,8 +268,13 @@ ssize_t sm2_file_open_or_create(struct sm2_mmap *map_shared)
 		}
 
 		common_fd = open(SM2_COORDINATION_FILE, O_RDWR);
-		if (common_fd > 0) {
+		if (common_fd >= 0) {
```
Comment (a-szegel) [inline prov/sm2/src/sm2_coordination.c:271]: This looks like a separate bugfix... plz put in different commit.

File: prov/sm2/src/sm2_coordination.c (hunk 2 above)
Comment (lrbison) [reply prov/sm2/src/sm2_coordination.c:195]: it's really: `sizeof(SM2_COORDINATION_DIR) + sizeof("/fi_sm2_pid1234567_XXXXXX") + 1` Which IIRC was something like 27.

File: prov/sm2/src/sm2_coordination.c (hunk 1 above)
Comment (lrbison) [reply prov/sm2/src/sm2_coordination.c:216]: we could, or we could let mmap will fail on an invalid fd, and return then.

File: prov/sm2/src/sm2_coordination.c (hunk 2 above)
Comment (lrbison) [reply prov/sm2/src/sm2_coordination.c]: addressed.

File: prov/sm2/src/sm2_coordination.c (hunk 1 above)
Comment (lrbison) [reply prov/sm2/src/sm2_coordination.c:216]: addressed.

File: prov/sm2/src/sm2_coordination.c (hunk 3 above)
Comment (lrbison) [reply prov/sm2/src/sm2_coordination.c:282]: 3 commits now.

**Review by wenduwan — APPROVED** [review]

**Review by a-szegel — APPROVED** [review]

File: prov/sm2/src/sm2_coordination.c [hunk 4]
```diff
@@ -268,8 +268,13 @@ ssize_t sm2_file_open_or_create(struct sm2_mmap *map_shared)
 		}
 
 		common_fd = open(SM2_COORDINATION_FILE, O_RDWR);
-		if (common_fd > 0) {
```
Comment (shijin-aws) [inline prov/sm2/src/sm2_coordination.c:271]: it can be 0? I thought fd should be positive

File: prov/sm2/src/sm2_coordination.c [hunk 5]
```diff
@@ -237,8 +244,15 @@ ssize_t sm2_file_open_or_create(struct sm2_mmap *map_shared)
 	 * data exchange regions yet.
 	 */
 	err = sm2_mmap_remap(&map_ours, header->ep_regions_offset);
-	if (err)
-		return err;
+
+	if (err) {
+	early_exit:
```
Comment (aingerson) [inline prov/sm2/src/sm2_coordination.c:250]: Don't love the look of this error handling. Can you modify to have a normal goto err and have the err case at the bottom with a goto err here instead? Looks kind of weird with this goto in the middle of the function and inside the if. Personal prefernce

File: prov/sm2/src/sm2_coordination.c (hunk 4 above)
Comment (lrbison) [reply prov/sm2/src/sm2_coordination.c:271]: lol, so did I, which is why I wrote it that way in the first place.
> The return value of open() is a file descriptor, a small,
       nonnegative integer that is an index to an entry in the process's
       table of open file descriptors.

File: prov/sm2/src/sm2_coordination.c (hunk 5 above)
Comment (lrbison) [reply prov/sm2/src/sm2_coordination.c]: I moved it to the bottom.  You are right, looks better.

File: prov/sm2/src/sm2_coordination.c [hunk 6]
```diff
@@ -268,8 +268,13 @@ ssize_t sm2_file_open_or_create(struct sm2_mmap *map_shared)
 		}
 
 		common_fd = open(SM2_COORDINATION_FILE, O_RDWR);
-		if (common_fd > 0) {
+		if (common_fd >= 0) {
 			tmp_header = sm2_mmap_map(common_fd, map_shared);
+			if (!tmp_header) {
+				/* failed to mmap. */
```
[the remark this answers is not in the document]
Comment (lrbison) [reply prov/sm2/src/sm2_coordination.c]: removed.

**Review by aingerson — APPROVED** [review]

**Review by shijin-aws — APPROVED** [review]
